pkg/bucketeer/sdk.go

```diff
@@ -1,87 +1,256 @@
+	evaluation, err := s.getEvaluation(ctx, user, featureID)
+	if err != nil {
+		s.logVariationError(err, "Float64Variation", user.Id, featureID)
+		s.eventProcessor.PushDefaultEvaluationEvent(ctx, user.User, featureID)
+		return defaultValue
+	}
+	variation := evaluation.Variation.Value
+	v, err := strconv.ParseFloat(variation, 64)
+	if err != nil {
+		s.logVariationError(err, "Float64Variation", user.Id, featureID)
+		s.eventProcessor.PushDefaultEvaluationEvent(ctx, user.User, featureID)
+		return defaultValue
+	}
+	s.eventProcessor.PushEvaluationEvent(ctx, user.User, evaluation)
+	return v
 }
 
-func (s *sdk) IntVariation(ctx context.Context, featureID string, user *User, defaultValue int) int {
-	return 0
+func (s *sdk) StringVariation(ctx context.Context, user *User, featureID, defaultValue string) string {
+	evaluation, err := s.getEvaluation(ctx, user, featureID)
+	if err != nil {
+		s.logVariationError(err, "StringVariation", user.Id, featureID)
+		s.eventProcessor.PushDefaultEvaluationEvent(ctx, user.User, featureID)
+		return defaultValue
+	}
+	variation := evaluation.Variation.Value
+	s.eventProcessor.PushEvaluationEvent(ctx, user.User, evaluation)
+	return variation
 }
 
-func (s *sdk) Int64Variation(ctx context.Context, featureID string, user *User, defaultValue int64) int64 {
-	return 0
+func (s *sdk) JSONVariation(ctx context.Context, user *User, featureID string, dst interface{}) {
+	evaluation, err := s.getEvaluation(ctx, user, featureID)
+	if err != nil {
+		s.logVariationError(err, "JSONVariation", user.Id, featureID)
+		s.eventProcessor.PushDefaultEvaluationEvent(ctx, user.User, featureID)
+		return
+	}
+	variation := evaluation.Variation.Value
+	err = json.Unmarshal([]byte(variation), dst)
+	if err != nil {
+		s.logVariationError(err, "JSONVariation", user.Id, featureID)
+		s.eventProcessor.PushDefaultEvaluationEvent(ctx, user.User, featureID)
+		return
+	}
+	s.eventProcessor.PushEvaluationEvent(ctx, user.User, evaluation)
 }
 
-func (s *sdk) Float64Variation(ctx context.Context, featureID string, user *User, defaultValue float64) float64 {
-	return 0.0
+func (s *sdk) getEvaluation(ctx context.Context, user *User, featureID string) (*protofeature.Evaluation, error) {
+	req := &protogateway.GetEvaluationsRequest{
+		Tag:       s.tag,
+		User:      user.User,
+		FeatureId: featureID,
+	}
+	reqStart := time.Now()
+	res, err := s.apiClient.GetEvaluations(ctx, req)
+	if err != nil || res == nil {
+		if status.Code(err) == codes.DeadlineExceeded {
+			s.eventProcessor.PushTimeoutErrorCountMetricsEvent(ctx, s.tag)
+		} else {
+			s.eventProcessor.PushInternalErrorCountMetricsEvent(ctx, s.tag)
+		}
+		return nil, fmt.Errorf("failed to get evaluation: %w", err)
+	}
+	s.eventProcessor.PushGetEvaluationLatencyMetricsEvent(ctx, time.Since(reqStart), s.tag, res.State.String())
+	s.eventProcessor.PushGetEvaluationSizeMetricsEvent(ctx, proto.Size(res), s.tag, res.State.String())
```

Comment:
To speed up the response, we could send these events `PushGetEvaluationLatencyMetricsEvent
` and `PushGetEvaluationSizeMetricsEvent` asynchronously like we do in the client SDK. 
What do you think?

Reply:
I think the speed is no problem because eventProcessor pushes these events to the in-memory queue.

Reply:
Ok, let's see how it goes when you run the tests.